File: app/src/main/java/ai/elimu/vitabu/ui/StoryBooksActivity.java
package ai.elimu.vitabu.ui;

import android.content.Intent;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.GridLayout;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import java.util.List;

import ai.elimu.analytics.utils.LearningEventUtil;
import ai.elimu.content_provider.utils.ContentProviderHelper;
import ai.elimu.model.enums.ReadingLevel;
import ai.elimu.model.enums.analytics.LearningEventType;
import ai.elimu.model.v2.gson.content.ImageGson;
import ai.elimu.model.v2.gson.content.StoryBookGson;
import ai.elimu.vitabu.BaseApplication;
import ai.elimu.vitabu.BuildConfig;
import ai.elimu.vitabu.R;
import ai.elimu.vitabu.ui.storybook.StoryBookActivity;
import ai.elimu.vitabu.util.SingleClickListener;

import static ai.elimu.vitabu.util.BitmapDecoder.decodeSampledBitmap;

public class StoryBooksActivity extends AppCompatActivity {

    final static String PICTURES_PATH = Environment.getExternalStorageDirectory() +
            "/Android/data/" +
            BuildConfig.CONTENT_PROVIDER_APPLICATION_ID +
            "/files/" + Environment.DIRECTORY_PICTURES + "/";

    private GridLayout storyBooksGridLayout;
    private ProgressBar storyBooksProgressBar;

    private List<StoryBookGson> storyBooks;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Log.i(getClass().getName(), "onCreate");
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_storybooks);

        storyBooksGridLayout = findViewById(R.id.storyBooksGridLayout);
        storyBooksProgressBar = findViewById(R.id.storybooks_progress_bar);

        // Fetch StoryBooks from the elimu.ai Content Provider (see https://github.com/elimu-ai/content-provider)
        storyBooks = ContentProviderHelper.getStoryBookGsons(getApplicationContext(), BuildConfig.CONTENT_PROVIDER_APPLICATION_ID);
        Log.i(getClass().getName(), "storyBooks.size(): " + storyBooks.size());
    }

    @Override
    protected void onStart() {
        Log.i(getClass().getName(), "onStart");
        super.onStart();

        // Reset the state of the GridLayout
        storyBooksProgressBar.setVisibility(View.VISIBLE);
        storyBooksGridLayout.setVisibility(View.GONE);
        storyBooksGridLayout.removeAllViews();

        ((BaseApplication) getApplication()).getExecutor().execute(new Runnable() {
            @Override
            public void run() {

                StoryBookGson storyBook;
                ReadingLevel readingLevel;

                // Create a View for each StoryBook in the list
                for (int index = 0; index < storyBooks.size(); index++) {
                    storyBook = storyBooks.get(index);
                    readingLevel = storyBook.getReadingLevel();

                    Log.i(getClass().getName(), "storyBook.getId(): " + storyBook.getId());
                    Log.i(getClass().getName(), "storyBook.getTitle(): \"" + storyBook.getTitle() + "\"");
                    Log.i(getClass().getName(), "storyBook.getDescription(): \"" + storyBook.getDescription() + "\"");

                    if (index == 0 || readingLevel != storyBooks.get(index - 1).getReadingLevel()) {
                        final View levelLayout = LayoutInflater.from(StoryBooksActivity.this).inflate(R.layout.activity_storybooks_level, storyBooksGridLayout, false);

                        GridLayout.LayoutParams layoutParams = new GridLayout.LayoutParams();
                        layoutParams.columnSpec = GridLayout.spec(0, getResources().getInteger(R.integer.gridlayout_column_count));
                        levelLayout.setLayoutParams(layoutParams);

                        final TextView levelTextView = levelLayout.findViewById(R.id.levelName);
                        if (readingLevel == null) {
                            levelTextView.setVisibility(View.GONE);
                        } else {
                            levelTextView.setText(String.format(getResources().getString(R.string.level), readingLevel.ordinal()+1));
                        }

                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                storyBooksGridLayout.addView(levelLayout);
                                if (storyBooksGridLayout.getChildCount() == storyBooks.size()) {
                                    storyBooksProgressBar.setVisibility(View.GONE);
                                    storyBooksGridLayout.setVisibility(View.VISIBLE);
                                }
                            }
                        });
                    }

                    final View storyBookView = LayoutInflater.from(StoryBooksActivity.this).inflate(R.layout.activity_storybooks_cover_view, storyBooksGridLayout, false);

                    // Fetch Image from the elimu.ai Content Provider (see https://github.com/elimu-ai/content-provider)
                    Log.i(getClass().getName(), "storyBook.getCoverImage(): " + storyBook.getCoverImage());
                    final ImageGson coverImage = ContentProviderHelper.getImageGson(storyBook.getCoverImage().getId(), getApplicationContext(), BuildConfig.CONTENT_PROVIDER_APPLICATION_ID);
                    if (coverImage != null) {
                        final ImageView coverImageView = storyBookView.findViewById(R.id.coverImageView);
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                String imageFilePath = PICTURES_PATH + coverImage.getId() + "_r" + coverImage.getRevisionNumber() + "." + coverImage.getImageFormat().toString().toLowerCase();
                                Log.i(getClass().getName(), "imageFile: " + imageFilePath);
                                coverImageView.setImageBitmap(decodeSampledBitmap(imageFilePath, 100, 100));
                            }
                        });
                    }

                    TextView coverTitleTextView = storyBookView.findViewById(R.id.coverTitleTextView);
                    coverTitleTextView.setText(storyBook.getTitle());

                    final StoryBookGson finalStoryBook = storyBook;
                    storyBookView.setOnClickListener(new SingleClickListener() {
                        @Override
                        public void onSingleClick(View v) {
                            Log.i(getClass().getName(), "onClick");

                            Log.i(getClass().getName(), "storyBook.getId(): " + finalStoryBook.getId());
                            Log.i(getClass().getName(), "storyBook.getTitle(): " + finalStoryBook.getTitle());

                            // Report learning event to the Analytics application (https://github.com/elimu-ai/analytics)
                            LearningEventUtil.reportStoryBookLearningEvent(finalStoryBook, LearningEventType.STORYBOOK_OPENED, getApplicationContext(), BuildConfig.ANALYTICS_APPLICATION_ID);

                            Intent intent = new Intent(getApplicationContext(), StoryBookActivity.class);
                            intent.putExtra(StoryBookActivity.EXTRA_KEY_STORYBOOK_ID, finalStoryBook.getId());
                            startActivity(intent);
                        }
                    });

                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            storyBooksGridLayout.addView(storyBookView);
                            if (storyBooksGridLayout.getChildCount() == storyBooks.size()) {
                                storyBooksProgressBar.setVisibility(View.GONE);
                                storyBooksGridLayout.setVisibility(View.VISIBLE);
                            }
                        }
                    });
                }
            }
        });
    }
}

File: app/src/main/java/ai/elimu/vitabu/ui/StoryBooksActivity.kt
package ai.elimu.vitabu.ui

import ai.elimu.analytics.utils.LearningEventUtil
import ai.elimu.common.utils.ui.setLightStatusBar
import ai.elimu.common.utils.ui.setStatusBarColorCompat
import ai.elimu.content_provider.utils.ContentProviderUtil
import ai.elimu.model.v2.enums.ReadingLevel
import ai.elimu.model.v2.enums.analytics.LearningEventType
import ai.elimu.model.v2.gson.content.StoryBookGson
import ai.elimu.vitabu.BuildConfig
import ai.elimu.vitabu.R
import ai.elimu.vitabu.databinding.ActivityStorybooksBinding
import ai.elimu.vitabu.databinding.ActivityStorybooksCoverViewBinding
import ai.elimu.vitabu.ui.storybook.StoryBookActivity
import ai.elimu.vitabu.util.SingleClickListener
import ai.elimu.vitabu.util.readImageBytes
import ai.elimu.vitabu.viewmodel.LoadStoryBooksUiState
import ai.elimu.vitabu.viewmodel.StoryBookViewModel
import ai.elimu.vitabu.viewmodel.StoryBookViewModelImpl
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.GridLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.bumptech.glide.Glide
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

@AndroidEntryPoint
class StoryBooksActivity : AppCompatActivity() {

    private lateinit var storyBookViewModel: StoryBookViewModel

    private val TAG: String = StoryBooksActivity::class.java.name
    private lateinit var binding: ActivityStorybooksBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.i(TAG, "onCreate")
        super.onCreate(savedInstanceState)

        binding = ActivityStorybooksBinding.inflate(layoutInflater)
        setContentView(binding.root)

        initViewModels()

        // Fetch StoryBooks from the elimu.ai Content Provider (see https://github.com/elimu-ai/content-provider)
        initData()
        window.apply {
            setLightStatusBar()
            setStatusBarColorCompat(R.color.colorPrimaryDark)
        }
    }

    override fun onStart() {
        Log.i(TAG, "onStart")
        super.onStart()

        // Reset the state of the GridLayout
        binding.storybooksProgressBar.visibility = View.VISIBLE
        binding.storyBooksGridLayout.visibility = View.GONE
        binding.storyBooksGridLayout.removeAllViews()
    }

    private fun initViewModels() {
        storyBookViewModel = ViewModelProvider(this)[StoryBookViewModelImpl::class.java]
    }

    private fun initData() {
        storyBookViewModel.getAllStoryBooks()

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                storyBookViewModel.uiState.collect { uiState ->
                    Log.d(TAG, "uiState collected uiState: $uiState")
                    when (uiState) {
                        is LoadStoryBooksUiState.Loading ->
                            binding.storybooksProgressBar.visibility = View.VISIBLE

                        is LoadStoryBooksUiState.Success -> {
                            binding.storybooksProgressBar.visibility = View.GONE
                            showStoryBooks(uiState.storyBooks)
                        }
                    }
                }
            }
        }
    }

    private fun showStoryBooks(storyBooks: List<StoryBookGson>) {
        CoroutineScope(Dispatchers.IO).launch {
            var storyBook: StoryBookGson
            var readingLevel: ReadingLevel?

            // Create a View for each StoryBook in the list
            for (index in storyBooks.indices) {
                storyBook = storyBooks[index]
                readingLevel = storyBook.readingLevel

                Log.i(TAG, "storyBook.getId(): " + storyBook.id)
                Log.i(TAG, "storyBook.getTitle(): \"" + storyBook.title + "\"")
                Log.i(TAG, "storyBook.getDescription(): \"" + storyBook.description + "\"")

                if (index == 0 || readingLevel != storyBooks[index - 1].readingLevel) {
                    val levelLayout = LayoutInflater.from(this@StoryBooksActivity).inflate(
                        R.layout.activity_storybooks_level,
                        binding.storyBooksGridLayout,
                        false
                    )

                    val layoutParams = GridLayout.LayoutParams()
                    layoutParams.columnSpec = GridLayout.spec(
                        0,
                        resources.getInteger(R.integer.gridlayout_column_count)
                    )
                    levelLayout.layoutParams = layoutParams

                    val levelTextView = levelLayout.findViewById<TextView>(R.id.levelName)
                    if (readingLevel == null) {
                        levelTextView.visibility = View.GONE
                    } else {
                        levelTextView.text =
                            String.format(
                                resources.getString(R.string.level),
                                readingLevel.ordinal + 1
                            )
                    }

                    withContext(Dispatchers.Main) {
                        binding.storyBooksGridLayout.addView(levelLayout)
                        if (binding.storyBooksGridLayout.childCount == storyBooks.size) {
                            binding.storybooksProgressBar.visibility = View.GONE
                            binding.storyBooksGridLayout.visibility = View.VISIBLE
                        }
                    }
                }

                val storyBookViewBinding = ActivityStorybooksCoverViewBinding.inflate(layoutInflater, binding.storyBooksGridLayout, false)

                // Fetch Image from the elimu.ai Content Provider (see https://github.com/elimu-ai/content-provider)
                Log.i(TAG,
                    "storyBook.getCoverImage(): " + storyBook.coverImage
                            + ". Id: " + storyBook.coverImage.id)
                val coverImage = ContentProviderUtil.getImageGson(
                    storyBook.coverImage.id,
                    applicationContext, BuildConfig.CONTENT_PROVIDER_APPLICATION_ID
                )

                if (coverImage != null) {
                    val coverImageView = storyBookViewBinding.coverImageView
                    CoroutineScope(Dispatchers.IO).launch {
                        readImageBytes(coverImage.id)?.let { bytes ->
                            withContext(Dispatchers.Main) {
                                Glide.with(this@StoryBooksActivity).load(bytes)
                                    .into(coverImageView)
                            }
                        }
                    }
                } else {
                    Log.w(TAG, "coverImage is null. Id: " + storyBook.coverImage.id)
                }

                val coverTitleTextView = storyBookViewBinding.coverTitleTextView
                coverTitleTextView.text = storyBook.title

                val finalStoryBook = storyBook
                storyBookViewBinding.root.setOnClickListener(object : SingleClickListener() {
                    override fun onSingleClick(v: View?) {
                        Log.i(TAG, "onClick")

                        Log.i(TAG, "storyBook.getId(): " + finalStoryBook.id)
                        Log.i(TAG, "storyBook.getTitle(): " + finalStoryBook.title)
                        Log.i(TAG,
                            "storyBook.getDescription(): " + finalStoryBook.description)

                        // Report learning event to the Analytics application (https://github.com/elimu-ai/analytics)
                        LearningEventUtil.reportStoryBookLearningEvent(
                            finalStoryBook, LearningEventType.STORYBOOK_OPENED,
                            applicationContext, BuildConfig.ANALYTICS_APPLICATION_ID
                        )

                        val intent = Intent(applicationContext, StoryBookActivity::class.java)
                        intent.putExtra(
                            StoryBookActivity.EXTRA_KEY_STORYBOOK_ID,
                            finalStoryBook.id
                        )
                        intent.putExtra(
                            StoryBookActivity.EXTRA_KEY_STORYBOOK_LEVEL,
                            finalStoryBook.readingLevel
                        )
                        intent.putExtra(
                            StoryBookActivity.EXTRA_KEY_STORYBOOK_DESCRIPTION,
                            finalStoryBook.description
                        )
                        startActivity(intent)
                    }
                })

                withContext(Dispatchers.Main) {
                    binding.storyBooksGridLayout.addView(storyBookViewBinding.root)
                    if (binding.storyBooksGridLayout.childCount == storyBooks.size) {
                        binding.storybooksProgressBar.visibility = View.GONE
                        binding.storyBooksGridLayout.visibility = View.VISIBLE
                    }
                }
            }
        }
    }
}
